runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -1014,41 +984,45 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 	}
 
 	/**
-	 * Asserts that the dataAddr field of the indexable object is correct.
+	 * Checks that the dataAddr field of the indexable object is correct.
 	 *
 	 * @param arrayPtr      Pointer to the indexable object
+	 * @return if the dataAddr field of the indexable object is correct
 	 */
 	MMINLINE bool
-	isCorrectDataAddr(J9IndexableObject *arrayPtr)
+	isValidDataAddr(J9IndexableObject *arrayPtr)
 	{
+		bool isValidDataAddress = true;
 
-		return (InlineContiguous == getArrayLayout(arrayPtr))
-			? isCorrectDataAddrForContiguousArraylet(arrayPtr)
-			: isCorrectDataAddrForDiscontiguousArraylet(arrayPtr);
-	}
+		if (_isIndexableDataAddrPresent) {
+			void *dataAddr = getDataAddrForIndexableObject(arrayPtr);
+			isValidDataAddress = isValidDataAddr(arrayPtr, dataAddr);
+		}
 
-	/**
-	 * Asserts that the dataAddr field of the contiguous arraylet is correct.
-	 *
-	 * @param arrayPtr      Pointer to the indexable object
-	 */
-	MMINLINE bool
-	isCorrectDataAddrForContiguousArraylet(J9IndexableObject *arrayPtr)
-	{
-		void *dataAddr = getDataAddrForContiguous(arrayPtr);
-		return isCorrectDataAddrContiguous(arrayPtr, dataAddr);
+		return isValidDataAddress;
 	}
 
 	/**
-	 * Asserts that the dataAddr field of the indexable object is correct.
+	 * Checks that the dataAddr field of the indexable object is correct.
 	 *
 	 * @param arrayPtr      Pointer to the indexable object
+	 * @return if the dataAddr field of the indexable object is correct
 	 */
 	MMINLINE bool
-	isCorrectDataAddrForDiscontiguousArraylet(J9IndexableObject *arrayPtr)
+	isValidDataAddr(J9IndexableObject *arrayPtr, void *dataAddr)
 	{
-		void *dataAddr = getDataAddrForDiscontiguous(arrayPtr);
-		return isCorrectDataAddrDiscontiguous(arrayPtr, dataAddr);
+		bool isValidDataAddress = false;
+		uintptr_t dataSizeInBytes = getDataSizeInBytes(arrayPtr);
+
+		if (0 == dataSizeInBytes) {
+			isValidDataAddress = ((dataAddr == NULL) || (dataAddr == (void *)((uintptr_t)arrayPtr + discontiguousIndexableHeaderSize())));
+		} else if (dataSizeInBytes < _omrVM->_arrayletLeafSize) {
+			isValidDataAddress = (dataAddr == (void *)((uintptr_t)arrayPtr + contiguousIndexableHeaderSize()));
+		} else {
+			isValidDataAddress = (dataAddr == NULL);
```

Comment:
this is ok for now, but once associated JIT changes are in, we should update this: at least disallow NULL, but if possible, check that it fits somewhere in off-heap range and that it's aligned properly (to off-heap regions)